pkg/providerid/providerid.go

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache v2.0 license.
+
+package providerid
+
+import (
+	"fmt"
+	"strings"
+)
+
+const (
+	ProviderName     string = "moc"
+	ProviderIDPrefix string = ProviderName + "://"
+)
+
+type HostType string
+
+const (
+	HostTypeVM        HostType = "vm"
+	HostTypeBareMetal HostType = "baremetal"
+)
+
+func FormatInstanceID(hostType HostType, machineName string) string {
+	switch hostType {
+	case HostTypeVM:
+		// Don't append the host type for VMs to maintain consistency with previous versions.
+		return machineName
+
+	default:
+		return string(hostType) + "/" + machineName
+	}
+}
+
+func FormatProviderID(hostType HostType, machineName string) string {
+	return ProviderIDPrefix + FormatInstanceID(hostType, machineName)
+}
+
+func ParseProviderID(providerID string) (HostType, string, error) {
+	if providerID == "" {
+		return "", "", fmt.Errorf("providerID is empty")
+	}
+
+	withoutPrefix := strings.TrimPrefix(providerID, ProviderIDPrefix)
```

Comment:
Would it make sense to use built in URL parsing logic?

https://golang.org/pkg/net/url/#Parse

Reply:
Probably not. While this looks like a URL, it doesn't really follow the URL rules.